### PR TITLE
Catalog: Send error responses in more edge cases

### PR DESCRIPTION
**Pull request by divergentdave**
While I was working on an UPDATE test, I noticed another case where we were failing to send a response in certain error conditions (leading to a test hang). This is related to #2572. I went through `Catalog::handle_request()` and its call tree to send error responses in more cases; we should have complete coverage now. Error responses were primarily missing in `Catalog::update()`, in fall-through code where no zone handler in a chain handled a request, and during response signing where `?` was used to return early. I changed the return type of some functions or methods from a `Result` to a `ResponseInfo`, which helped flush out the latter. I also changed `MessageResponse` so that it only holds a reference to the response's `Edns` struct, not ownership of it, in order to eliminate some unnecessary clones.

### CODE REVIEW

File: crates/server/src/zone_handler/catalog.rs [hunk 1]
```diff
@@ -651,6 +649,25 @@ async fn zone_transfer(
     Err(LookupError::ResponseCode(ResponseCode::ServFail))
 }
 
+/// Helper function to construct a response message with an error response code, and send it via a
+/// response handler.
+async fn send_error_response(
+    response_code: ResponseCode,
+    request: &Request,
+    response_edns: Option<Edns>,
+    mut response_handle: impl ResponseHandler,
```
Comment (cpu) [inline crates/server/src/zone_handler/catalog.rs]: nit: it feels like having `request` as the first parameter, and then `response_code`, `response_edns`, and `response_hande` is a touch more natural. WDYT?

**Review by cpu — APPROVED** [review]
Nice! I think a few of those mistaken early returns were mine, thanks for spotting/fixing up.
